### PR TITLE
H383: Test-Time Training via Surface kNN Distance Self-Supervised Pretext

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Test-Time Training (TTT) via Surface kNN Distance Self-Supervised Pretext** — at inference time, for each val/test case, do K=3 small gradient steps on a GT-free self-supervised pretext task (predict surface kNN-8 mean distance per point from input coords), then forward through the adapted model to predict τ_z. Weights restored between cases.

**Why this is the right next probe (post-29-axis closure)**:

The consolidated meta-finding from 29 closed axes is:
- **meta-α**: ANY perturbation to the EP15 EMA basin under 3-epoch Lion cosine tail produces uniform multi-channel regression (H375 cross-channel decoder + H379 PEFT-LR both confirm, 7-channel uniform +1.1–1.7bp regression).
- **meta-β**: ANY case-distribution intervention is bottlenecked by the near-uniform per-case difficulty signal (H378 hard-case-mining confirms; w_max/w_min ≈ 1.36×, ESS ≈ 0.998).

Both meta-findings say: training-time perturbation is exhausted. The plateau-protocol implication is that **inference-time methods (TTT, per-case adaptation, post-hoc inference-time backprop) are the primary untapped attack vector**.

TTT operates on per-case parameter neighborhoods — each test case sees a tiny adaptation away from the EP15 EMA basin, then the adaptation is discarded. The shared training-time weights are never modified. This sidesteps **meta-α** entirely (no basin perturbation at training time) and is orthogonal to **meta-β** (no case-distribution reweighting, just per-case adaptation given the case is already at inference time).

**Mechanistic prediction**: if the encoder has any unused capacity to specialize features for a given case's local surface density pattern, K=3 pretext steps on a kNN-distance prediction task should activate that capacity. The downstream τ_z prediction should then reflect a slightly better-conditioned encoder representation for that case. If the encoder is fully saturated at inference time (no unused capacity), TTT will null cleanly without basin disruption.

**Literature context**: TTT-MAE (Sun et al. 2020, "Test-Time Training with Self-Supervision for Generalization under Distribution Shifts") established TTT as a viable inference-time adaptation method on classification benchmarks. For CFD surrogates on unseen geometries, the same per-case "domain-shift-like" intuition applies — each test case has a unique surface point distribution.

### Phase 0: Inspect the existing inference pipeline

Before implementing, study `target/train.py` and identify:
1. Where the eval/test loop iterates over cases (likely a `def evaluate(...)` or similar)
2. How EP15 EMA weights are currently loaded for eval (search for `ema_model` references)
3. How K=4 antithetic TTA is currently applied (likely an inner loop over `K` perturbed inputs)
4. The exact CLI flags for the TTA-only path (the `--mode evaluate` style invocation)

Post findings as a follow-up comment before writing TTT code. We want one source of truth on how the current eval pipeline is structured.

### Phase 1: Implement TTT with K=3 pretext steps

1. **Pretext head** (new module):
   ```python
   class PretextHead(nn.Module):
       def __init__(self, in_dim=64, hidden=32):
           super().__init__()
           self.fc1 = nn.Linear(in_dim, hidden)
           self.fc2 = nn.Linear(hidden, 1)
           nn.init.zeros_(self.fc2.weight)
           nn.init.zeros_(self.fc2.bias)
       def forward(self, x):
           return self.fc2(F.relu(self.fc1(x)))
   ```
   `in_dim` should match the encoder's per-point latent dim. Verify before instantiating.

2. **Pretext target** (computed on-the-fly from input coords, no GT needed):
   ```python
   def knn8_mean_dist(coords, k=8):
       # coords: [N_points, 3]
       d = torch.cdist(coords, coords)            # [N, N]
       d, _ = torch.topk(d, k=k+1, largest=False) # [N, k+1] (includes self at dist=0)
       return d[:, 1:].mean(dim=-1, keepdim=True) # [N, 1] — drop self
   ```
   Compute once per case and detach.

3. **TTT per-case loop** (single GPU; for DDP see Phase 1.5):
   ```python
   ema_state = {k: v.clone() for k, v in ema_model.state_dict().items()}
   pretext_head = PretextHead().cuda()  # NEW module, not in EMA snapshot

   for case in val_test_dataset:
       # Restore EP15 EMA before each case
       model.load_state_dict(ema_state)
       # Re-init pretext head (zero output)
       pretext_head = PretextHead().cuda()
       optimizer = AdamW(
           list(model.parameters()) + list(pretext_head.parameters()),
           lr=1e-5, betas=(0.9, 0.999), weight_decay=0.0
       )

       # K=3 pretext steps
       model.train()  # but keep DROPOUT_OFF — use eval-style dropout if any
       d_target = knn8_mean_dist(case.coords).detach()
       for step in range(K):
           z_local = model.encoder(case.coords, case.features)  # [N, 64]
           d_hat = pretext_head(z_local)
           pretext_loss = F.mse_loss(d_hat, d_target)
           pretext_loss.backward()
           optimizer.step()
           optimizer.zero_grad()

       # Final prediction with adapted weights
       model.eval()
       with torch.no_grad():
           y_pred = model(case.input)  # this is where K=4 antithetic TTA gets applied as usual
       # Store y_pred for case
   ```

4. **Updated params during TTT**:
   - **Update**: backbone encoder + slice-attention layers + pretext_head
   - **Freeze**: surface decoder (closest to output, riskier to perturb at inference)
   - Set `requires_grad=False` on all `surface_decoder.*` params before constructing optimizer.

5. **Key invariant** (verify in smoke): the EP15 EMA snapshot must be bit-exact restored between cases — confirm via `torch.allclose(model.state_dict()[k], ema_state[k])` for a few k after `load_state_dict()`.

### Phase 1.5: DDP considerations

Run TTT under DDP=8 with **per-rank case partitioning**:
- Each rank processes its own subset of cases (e.g. test=168 → 21 cases/rank)
- DDP gradient sync **disabled** inside the TTT inner loop (`with model.no_sync():` or set `find_unused_parameters=False` and call `model.module` directly for the per-case loop)
- Each rank does its own per-case TTT independently — no cross-rank gradient sync needed
- After TTT, gather final predictions from all ranks via `dist.all_gather_object()` or similar
- This keeps TTT correct (per-case independence preserved) and fast (full DDP parallelism)

### Phase 2: Smoke check (≤ 30 min, ≤ 5 cases)

Run TTT on the first 5 val cases. Verify:
1. **Pretext loss decreases ≥30% across K=3 steps** (confirms gradient signal exists). If pretext loss is flat or increases, the pretext task has no signal → CLOSE.
2. **Predictions are numerically different** from EP15 EMA baseline (rmsd of per-case y_pred > 1e-5, < 1e-2) — TTT actually changes outputs.
3. **No NaN/Inf** in adapted weights or predictions.
4. **EP15 EMA restoration is bit-exact** — `state_dict()` matches `ema_state` after `load_state_dict()`.
5. **Implementation runs cleanly under DDP=8** for at least 2 ranks × 1 case each.

Smoke results table:
| Case | Pretext step 0 | Pretext step 3 | Δ | y_pred rmsd vs EMA |
|---|---:|---:|---:|---:|
| ... | ... | ... | ... | ... |

### Phase 3: Full val+test pass with TTT + cal

Run TTT on all 40 val + 168 test cases (per-rank partitioned). Apply existing K=4 antithetic TTA on top of the per-case-adapted model. Compute val_RAW / val_cal / test_cal using existing H336 cal coefs.

**Final metrics** (compared against H342 SOTA baseline):

| Metric | H342 baseline | H383 TTT result | Δ | Verdict |
|---|---:|---:|---:|---|
| val_RAW abupt | (compute) | ... | ... | ... |
| val_cal abupt | **5.8962%** | ... | ... | ... |
| test_cal abupt | **5.7357%** | ... | ... | ... |
| test_WSS_z | **8.6122%** | ... | ... | (target Transolver-3 5.85%) |

## Kill gates (pre-committed)

- **Smoke gate**: pretext loss decrease < 10% across K=3 steps → **CLOSE** (pretext task has no signal on this model)
- **Phase 1 val gate (post-TTT, pre-cal)**: val_RAW abupt > 6.05% → **CLOSE** (TTT broke the basin even though we tried to be careful)
- **Phase 2 cal gate (final)**: val_cal abupt > 5.8962 → **CLOSE** (fails merge gate)
- **Phase 2 test gate (final)**: test_cal abupt > 5.7357 → **CLOSE** (fails merge gate)

If smoke fires (pretext null): post a clean Finding entry explaining the pretext task has no gradient signal at the current encoder capacity. The result is still informative for the inference-time-adaptation tier.

## Baseline (compare against)

H342 3-cp output-avg ep13+ep14+ep15 × K=4 TTA (PR #1526 MERGED 2026-06-01 19:15Z):
- val_cal abupt: **5.8962%**
- test_cal abupt: **5.7357%**
- test_WSS_z: **8.6122%**
- test_VP: **3.3751%**
- test_SP: **3.6124%**
- W&B runs: 3icmxaqe/qgw0ix77/ijadzof0

**Merge gate**: val_cal abupt < 5.8962 AND test_cal abupt < 5.7357 (AND-logic).

**Reproduce H342 baseline**:
```bash
cd "target/" && python train.py --mode evaluate --eval-mode three-cp-tta-k4 --tta-noise-family studentt --tta-noise-df 4 --tta-noise-sigma 1.0e-4 --tta-K 4 --three-cp-eps "13,14,15" --eval-resume <H342-ckpts> --wandb_group h383-edward-ttt
```
(Adjust flags to match the actual H342 reproduce command in `BASELINE.md`.)

**Reproduce H383 TTT** (Phase 3 final):
```bash
cd "target/" && python train.py --mode evaluate --eval-mode three-cp-tta-k4-ttt \
  --ttt-k 3 --ttt-lr 1e-5 --ttt-pretext knn8_mean_dist \
  --ttt-freeze surface_decoder \
  --tta-noise-family studentt --tta-noise-df 4 --tta-noise-sigma 1.0e-4 --tta-K 4 \
  --three-cp-eps "13,14,15" \
  --eval-resume <H342-ckpts> \
  --wandb_group h383-edward-ttt \
  --kill-thresholds "PHASE1:val_primary/abupt_axis_mean_rel_l2_pct<=6.05"
```

**DDP**: Use 8 GPUs (`torchrun --nproc-per-node=8 ...`) with per-rank case partitioning during TTT inner loop. Disable gradient sync during pretext gradient steps; sync only at final prediction gathering.

**Wallclock estimate**: 
- Implementation + smoke: 4-5h
- Phase 1 val with TTT: ~1h (40 cases × K=3 × ~10s/step × 8 GPUs in parallel = ~75s wall)
- Phase 3 test with TTT + cal: ~3h
- Total: ~7-9h (fits in 540min budget)

**SENPAI_TIMEOUT_MINUTES**=540 hard cap. **SENPAI_MAX_EPOCHS** N/A (no training, eval-only).

## Risks / pitfalls

1. **Per-case state restore must be bit-exact** — use in-memory state_dict copy, never disk save/restore (too slow). Verify via `torch.allclose` smoke check.
2. **TTT might trigger broad-disruption signature** if K too large or LR too high — K=3, LR=1e-5 are conservative. If smoke shows y_pred rmsd > 1e-2 from EMA baseline on the first case, abort and post diagnostic; do NOT proceed to full val.
3. **Pretext head must be re-initialized per case** — otherwise it accumulates pretext signal across cases, defeating per-case independence.
4. **DDP gradient sync** must be disabled in TTT inner loop. If your DDP wrapper auto-syncs on every backward(), use `with model.no_sync():` or manually call backward through `model.module`.
5. **Surface decoder frozen during TTT** — set `requires_grad=False` on `surface_decoder.*` params before constructing optimizer. Verify in smoke that pre/post-TTT decoder weights are identical.

## Notes

- Kill-threshold operator semantics (banked from fern H379): `STEP:METRIC<=VALUE` is the SURVIVAL condition encoding "kill if METRIC > VALUE". Use `<=` not `>`.
- This hypothesis is mechanistically distinct from all 29 closed axes. The closest mechanistic neighbor is H374 self-consistency (Finding X) — but that used a frozen EP15 EMA teacher with ~9% WSS_z error to supervise the student, whereas H383 uses a GT-free analytic pretext (kNN distance) and adapts only at inference time (not training time).
- If TTT nulls: confirms the encoder is fully saturated at inference time and inference-time per-case backprop cannot extract additional signal from the H336 basin. We then move to other inference-time tiers (variance-aware cal, multi-seed inference, ensemble disagreement features).
- If TTT wins (val_cal < 5.8962 AND test_cal < 5.7357): opens an entirely new attack vector for the post-29-axis plateau. We would explore K (1, 5, 7), pretext task variants (curvature, local PE, surface autoencode), and per-channel vs global TTT.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward — Phase 0 inspection:

## Existing eval pipeline (`eval_tta_h252.py`)

The PR's reproduce command references a fictional `train.py --mode evaluate` interface; the real flow lives in `eval_tta_h252.py` + `tools/h342_avg_checkpoints.py`. Findings:

1. **Eval loop**: `evaluate_split_stacked()` (eval_tta_h252.py:824) iterates over per-rank case subsets via `my_cases = case_ids[rank::world_size]`. Each case is processed by `process_case_stacked()` (eval_tta_h252.py:355).
2. **EMA loading**: Direct `torch.load(ckpt_path)` + `model.load_state_dict()` (line 1048). The checkpoint *is* the EMA snapshot already (best_checkpoint_source="ema" in train.py). Then `clean = snapshot_clean_params(model)` (line 1064) holds the in-memory restore source.
3. **K=4 antithetic TTA**: Schedule = `[(k, sign) for k in range(K_passes) for sign in (+1.0, -1.0)]` (line 405). Each pass perturbs `p ← clean + sign·draw(seed)·sigma·|clean|` via `perturb_relative_()` (line 320), evaluates across all (resolutions × mirror), then proceeds to the next k. Final restore of clean weights at line 1187.
4. **Per-case raw predictions saved** via `--save-raw-predictions <path>` (line 1192-1230), keyed by case_id, then composed downstream by `tools/h342_avg_checkpoints.py` for 3-cp output averaging.
5. **DDP**: `init_distributed()` at line 988 uses 120-min NCCL timeout (lazy comm init can wait >600s default). Final `dist.all_gather_object()` only at the end of each split (line 926).

## Model layout (`model.py`)

- Encoder block: `pos_embed`, `{surface,volume}_string_sep`, `{surface,volume}_bias`, `project_{surface,volume}_features`, `{surface,volume}_placeholder`.
- Backbone (slice-attention): `backbone.blocks[0..4]` (5 layers, hidden=512, n_head=4, num_slices=128).
- Norm: `norm` (LayerNorm).
- xattn: `surf_to_vol_xattn` + `surf_to_vol_xattn_norm`.
- **Decoders**: `surface_out` (Linear 512→512 → SiLU → Linear 512→4 for cp, τ_x, τ_y, τ_z) and `volume_out` (deeper 3-layer MLP).
- Per-point encoder latent dim is **n_hidden=512** (not 64 as in PR template). `surface_hidden` returned by forward at line 679 is [B, N, 512].

## Implementation plan (deviations from PR template)

- Add TTT as gated feature in `eval_tta_h252.py` (not a fork) — preserves all H336/H342 reproductions and produces `_raw.npz` files compatible with `tools/h342_avg_checkpoints.py` for 3-cp averaging.
- PretextHead with `in_dim=512, hidden=64` (n_hidden=512 here).
- Pretext target = chunked kNN-8 mean distance on a downsampled surface coord set (16384 points) to keep memory bounded.
- TTT forward uses **surface tokens only** (volume_x=None) to avoid disturbing volume_out / volume_bias / volume_string_sep paths. The backbone slice-attention is shared and gets updated.
- Freeze `surface_out.*` + `volume_out.*` (decoder); update everything else. Pretext head is per-case re-initialized (fc2 zero-init so initial output is zero).
- Per-rank case partitioning + DDP gradient sync disabled via direct `unwrap_model(model)` calls (already how existing H336/H342 path works — no DDP-wrapped model).

## Baseline & gates (corrected for current 34/50 split)

- H336 single-cp baseline (val 5.8978 / test 5.7379) is the realistic single-checkpoint hurdle.
- H342 3-cp baseline (val 5.8962 / test 5.7357) is what we need to beat for the merge gate. To clear that with TTT on a single checkpoint requires TTT to give ≥1.6bp val / ≥2.2bp test gain — non-trivial but possible.
- Plan: smoke on 5 val cases with EP15; if smoke passes, run full val+test with EP15; if EP15-only TTT shows positive signal at the H336 hurdle, expand to 3-cp (EP13+14+15 with TTT each) for the H342 merge gate.

Implementing TTT now.
